Fixing regex for kubernetes version in kubeadm

Pensu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
/kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Since https://github.com/kubernetes/test-infra/pull/14030 got merged, the marker for cross builds has moved from ci-cross/latest to ci/k8s-master. While providing the marker to kubeadm via --kubernetes-version flag, ci/k8s-master fails with:

```
version "ci/k8s-master" doesn't match patterns for neither semantic version nor labels
```

This PR fixes the release label regex so that kubeadm can recognize ci/k8s-master marker.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubeadm: add support for the "ci/k8s-master" version label as a replacement for "ci-cross/*", which no longer exists.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Pensu committed:
Fixing regex for kubernetes version in kubeadm

Pensu commented:
/test pull-kubernetes-bazel-test

mkumatag commented:
/cc @neolit123 @SataQiu @BenTheElder can you please help to review this patch?

mkumatag commented:
@Pensu I see few testcases are failing:

```
--- FAIL: TestCIBuildVersion (0.00s)
    --- FAIL: TestCIBuildVersion/input:release/0invalid/expected: (0.00s)
        version_test.go:289: Input: "release/0invalid". Result: "v1.17.0", Error: <nil>
        version_test.go:294: KubernetesReleaseVersion: error expected for input "release/0invalid", but result is "v1.17.0"
    --- FAIL: TestCIBuildVersion/input:ci/0invalid/expected: (0.00s)
        version_test.go:289: Input: "ci/0invalid". Result: "v1.17.0", Error: <nil>
        version_test.go:294: KubernetesReleaseVersion: error expected for input "ci/0invalid", but result is "v1.17.0"
```

Pensu commented:
@mkumatag Yeah, the reason is the current regex doesn't expect a numeric in the string before "-" i.e. it should be in the form of ci/latest, ci/latest-1.0, ci/stable, ci/stable-1.0 etc. Hence the failure. But the new marker for cross build has the number "8" (ci/k8s-master), so I believe this is expected. If I understand correctly, once done, the test cases would have to be updated as well.

neolit123 reviewed (commented):
adapting the regexp is fine.
but please fix the existing unit tests and add new ones to cover this new case.

neolit123 commented:
/priority important-soon

neolit123 commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -40,7 +40,7 @@ const (
 	kubeReleaseBucketURL  = "https://dl.k8s.io"
 	kubeReleaseRegex      = regexp.MustCompile(`^v?(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)([-0-9a-zA-Z_\.+]*)?$`)
-	kubeReleaseLabelRegex = regexp.MustCompile(`^[[:lower:]]+(-[-\w_\.]+)?$`)
+	kubeReleaseLabelRegex = regexp.MustCompile(`^[a-z0-9]+(-[-\w_\.]+)?$`)
```
both the old and new regexp are not optimal.

please use this:
https://play.golang.org/p/zprnyzsKDnV

and add a unit test for `k8s-master`.

neolit123 commented:
> version "ci/k8s-master" doesn't match patterns for neither semantic version nor labels

please change this release note to:
> kubeadm: add support for the "ci/k8s-master" version label

Pensu commented:
/retest

Pensu commented:
@neolit123 Thanks for the review. I have updated the PR with:

1) The regex you provided.
2) Changed the expected error value for invalid lower case strings. As now, we don't fall back to any version, we just throw an error.
3) Removed ci-cross references and added ci/k8s-master.
4) Added the release note.

PTAL.

neolit123 commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -238,9 +234,8 @@ func TestKubernetesIsCIVersion(t *testing.T) {
 		// CI builds
 		{"ci/latest-1", true},
-		{"ci-cross/latest", true},
+		{"ci/k8s-master", true},
```
for reference:
https://github.com/kubernetes/test-infra/pull/14030
https://github.com/kubernetes/release/issues/854#issuecomment-524071615

`k8s-master` on the other hand is a label that i don't fully understand.
if you look here there are more `k8s-*` labels but i don't think we should support them in kubeadm for now.
https://console.cloud.google.com/storage/browser/kubernetes-release-dev/ci?pli=1

so i'm fine with the replacement this PR is doing.

Pensu replied:
Yeah, I agree. k8s-master can be used to verify the cross builds for kubernetes when used with kubeadm.

neolit123 commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -269,9 +264,7 @@ func TestCIBuildVersion(t *testing.T) {
-		{"ci-cross/v1.9.0-alpha.1.123+acbcbfd53bfa0a", "v1.9.0-alpha.1.123+acbcbfd53bfa0a", true},
 		{"ci/1.9.0-alpha.1.123+acbcbfd53bfa0a", "v1.9.0-alpha.1.123+acbcbfd53bfa0a", true},
-		{"ci-cross/1.9.0-alpha.1.123+acbcbfd53bfa0a", "v1.9.0-alpha.1.123+acbcbfd53bfa0a", true},
 		{"ci/0invalid", "", false},
```
can we have `k8s-master` here?

Pensu replied:
This section tests release versions. However, afaiu k8s-master doesn't contain a specific version, but it keeps changing based on the latest cross builds available. So, it won't be possible to test k8s-master with a specific version. I guess that's the same reason why we don't test ci/latest in this particular section. Please correct me if my understanding is wrong here.

neolit123 replied:
SGTM.

neolit123 commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -113,17 +112,15 @@ func TestVersionFromNetwork(t *testing.T) {
-		"unknown":    {"The requested URL was not found on this server.", currentVersion, true, false},
+		"empty":      {"", "", true, true},
+		"garbage":    {"<?xml version='1.0'?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message></Error>", "", true, true},
+		"unknown":    {"The requested URL was not found on this server.", "", true, true},
```
i don't think these changes are valid.
`kubernetesReleaseVersion()` does fallback to the version in a constant:
https://github.com/kubernetes/kubernetes/blob/67d928acdc351d3d4f9e20a92cee4452a26ed0c4/cmd/kubeadm/app/util/version.go#L109

Pensu replied:
That case applies when we have been able to parse the version with `kubeReleaseLabelRegex.MatchString(versionLabel)` section but not able to find that version. With the change in the regex, we would only be parsing strings that start with 'k8s-master`, 'latest` and `stable`. For everything else (i.e. empty, garbage etc), the control will go to: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/version.go#L125 resulting in error. Please let me know if that assumption is not correct.

neolit123 replied:
what seems wrong here is that both `FetcherErrorExpected` and `ErrorExpected` now always match in the test cases. if the regex now catches the badly formatted labels, then perhaps it makes sense to only have a single `ErrorExpected`.

but we still need a test that validates the fallback to the constant (local version) in case the network is not available. that was the purpose of the `currentVersion` variable used as a test case result.

Pensu replied:
Yeah, sure, makes sense. I would leave these tests as is and will add another test, that will check the fall back flow and will return back to currenVersion. Does that sounds good?

neolit123 replied:
ok

neolit123 commented:
we probably should note `ci-cross` is no longer supported.

minor update to the release note:

> kubeadm: add support for the "ci/k8s-master" version label as a replacement for "ci-cross/*", which no longer exists.

Pensu commented:
> we probably should note `ci-cross` is no longer supported.
> 
> minor update to the release note:
> 
> > kubeadm: add support for the "ci/k8s-master" version label as a replacement for "ci-cross/*", which no longer exists.

@neolit123 Done.

neolit123 commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -121,9 +121,10 @@ func TestVersionFromNetwork(t *testing.T) {
-		"unknown":    {"The requested URL was not found on this server.", currentVersion, true, false},
+		"empty":      {"", "", true, true},
+		"garbage":    {"", "", true, true},
+		"unknown":    {"", "", true, true},
```
instead of these 3 that are essentially the same test case, IMO we should have a list of bad inputs to the regexp to keep it from breaking when changed in the future.

e.g. see my go playground tests:
https://play.golang.org/p/zprnyzsKDnV

neolit123 replied:
delete my previous comments, as they were wrong.
please format this like so and the PR is good to go:

```
	cases := map[string]T{
		"stable":          {"stable-1", "v1.4.6", false, false}, // recursive pointer to stable-1
		"stable-1":        {"v1.4.6", "v1.4.6", false, false},
		"stable-1.3":      {"v1.3.10", "v1.3.10", false, false},
		"latest":          {"v1.6.0-alpha.0", "v1.6.0-alpha.0", false, false},
		"latest-1.3":      {"v1.3.11-beta.0", "v1.3.11-beta.0", false, false},
		"latest-1.5":      {"", currentVersion, true, false}, // fallback to currentVersion on fetcher error
		"invalid-version": {"", "", false, true},             // invalid version cannot be parsed
	}
```

neolit123 commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -121,9 +121,10 @@ func TestVersionFromNetwork(t *testing.T) {
+		"empty":      {"", "", true, true},
+		"garbage":    {"", "", true, true},
+		"unknown":    {"", "", true, true},
+		"latest-1.5": {"", currentVersion, true, false},
```
but we also have to simulate a network failure with the `fileFetcher` that does the fallback to currentVersion.
let me know if you have further questions.

IMO this block has to be changed like so:
```
if v.FetcherErrorExpected {
    return currentVersion, errors.New("expected error")
}
```
because if there is a fetcher error the fallback is always currentVersion.

neolit123 reviewed (commented):
/lgtm
/approve
thanks for the update

Pensu commented:
/retest

neolit123 commented:
> pull-kubernetes-verify — Job failed.

@Pensu 
try calling `./hack/update-gofmt.sh`.

neolit123 commented:
/lgtm

neolit123 commented:
/retest

Pensu commented:
/retest

neolit123 commented:
/retest
